File: 72.edit-distance.py
class Solution:
    def minDistance(self, word1, word2):
        """
        :type word1: str
        :type word2: str
        :rtype: int
        """
        m = len(word1)
        n = len(word2)
        dp = [_ for _ in range(n + 1)]
        for i in range(1, m + 1):
            pre = dp[0]
            dp[0] = i
            for j in range(1, n + 1):
                cur = dp[j]
                if word1[i - 1] == word2[j - 1]:
                    dp[j] = pre
                else:
                    dp[j] = min(pre, dp[j], dp[j - 1]) + 1
                pre = cur
        return dp[n]

File: 72.edit-distance.cpp
class Solution {
public:
    int minDistance(string word1, string word2) {
        int m = word1.length();
        int n = word2.length();
        vector<int> dp(n + 1, 0);
        for (int i = 0; i <= n; i++) {
            dp[i] = i;
        }
        for (int i = 1; i <= m; i++) {
            int pre = dp[0];
            dp[0] = i;
            for (int j = 1; j <= n; j++) {
                int cur = dp[j];
                if (word1[i - 1] == word2[j - 1]) {
                    dp[j] = pre;
                } else {
                    dp[j] = min(pre, min(dp[j], dp[j - 1])) + 1;
                }
                pre = cur;
            }
        }
        return dp[n];
    }
};